internal/controllers/machine/machine_controller.go

```diff
@@ -791,38 +803,89 @@ func shouldRequeueDrain(now time.Time, lastDrain time.Time) (time.Duration, bool
 	if err := remoteClient.Get(ctx, types.NamespacedName{Name: nodeName}, node); err != nil {
 		if apierrors.IsNotFound(err) {
 			log.Error(err, "Could not find node from noderef, it may have already been deleted")
-			return false, nil
+			return ctrl.Result{}, nil
 		}
-		return true, err
+		return ctrl.Result{}, err
 	}
 
 	if noderefutil.IsNodeUnreachable(node) {
 		// If a node is unreachable, we can't detach the volume.
 		// We need to skip the detachment as we otherwise block deletions
 		// of unreachable nodes when a volume is attached.
 		log.Info("Node is unreachable, skip waiting for volume detachment.")
-		return false, nil
+		return ctrl.Result{}, nil
 	}
 
-	if len(node.Status.VolumesAttached) != 0 {
-		log.Info("Waiting for Node volumes to be detached")
-		return true, nil
+	// Check waitForVolumeDetach cache to ensure we won't retry before waitForVolumeDetachInterval.
+	// Note: This is intentionally only done if we have Pods to evict, because otherwise
+	// even the "no-op" drain would be requeued.
+	if cacheEntry, ok := r.waitForVolumeDetachCache.Has(client.ObjectKeyFromObject(machine)); ok {
+		if requeueAfter, requeue := shouldRequeue(time.Now(), cacheEntry.LastProceed, waitForVolumeDetachInterval); requeue {
+			log.Info(fmt.Sprintf("Requeuing in %s because there already was a wiat for node volumes in the last %s", requeueAfter, waitForVolumeDetachInterval))
+			return ctrl.Result{RequeueAfter: requeueAfter}, nil
+		}
 	}
 
-	return false, nil
+	// Get two sets of information about volumes currently attached to the node:
+	// * VolumesAttached names from node.Status.VolumesAttached
+	// * PersistentVolume names from VolumeAttachments with status.Attached set to true
+	attachedVolumeNames, attachedPVNames, err := getAttachedVolumeInformation(ctx, remoteClient, node)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Return early if there are no volumes to wait for getting detached.
+	if len(attachedVolumeNames) == 0 && len(attachedPVNames) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// Add entry an to the cache so we won't retry before waitForVolumeDetachInterval.
+	r.waitForVolumeDetachCache.Add(drain.CacheEntry{
+		Machine:     client.ObjectKeyFromObject(machine),
+		LastProceed: time.Now(),
+	})
+
+	// Get all PVCs we want to ignore because they have pods running which should skip drain.
+	pvcsToIgnoreFromPods, err := getPersistentVolumeClaimsToIgnore(ctx, remoteClient, nodeName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// List all PersistentVolumes and return the ones we want to wait for.
+	pvsWaitingForDetach, err := getPersistentVolumesWaitingForDetach(ctx, remoteClient, attachedVolumeNames, attachedPVNames, pvcsToIgnoreFromPods)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If no pvs were found we have to wait for and there are no unmatched VolumeHandles
+	// or PVs from VolumeAttachments, then we are finished with waiting for volumes.
+	if len(pvsWaitingForDetach) == 0 {
```

Comment:
Hm, I think even with moving the cache stuff around we would end up with the 20s requeue for the case where we only have volumes left that we want to ignore